test/WebSites/MvcTagHelpersWebSite/Views/MvcTagHelper_Home/ProductList.cshtml

```diff
@@ -10,9 +10,17 @@
 <body>
     @using (@Html.BeginForm("Index", "MvcTagHelper_Product", FormMethod.Post))
     {
-        @foreach (var product in Model)
+        var index = 0;
+        var fieldPrefix = ViewData.TemplateInfo.HtmlFieldPrefix;
+        foreach (var model in Model)
         {
-            @await Html.PartialAsync("_ProductPartial", product)
+            ViewData.TemplateInfo.HtmlFieldPrefix = fieldPrefix + string.Format("[{0}]", index++);
```

Comment:
Perhaps write a Razor comment here explaining why this is done. It's not the most obvious of pieces.